### PR TITLE
Configure JAAS settings for Kafka via Spring Boot

### LINKED ISSUE #36: Making AdminUtils and topic creation optional in Kafka binder
If an end user does not want to include the kafka broker jar in the classpath, we should support that model and not create the topic and simply ignore the fact that AdminUtils is not present. It shouldn't stop the binder.

### PR DESCRIPTION
**Pull request by mbogoevici**
Fixes #36

Adds support for programmatic configurtion of JAAS
properties as an alternative to using a JAAS file.

### CODE REVIEW

File: spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc [hunk 1]
```diff
@@ -215,18 +215,120 @@ spring.cloud.stream.kafka.binder.configuration.security.protocol=SASL_SSL
...
+spring.cloud.stream.kafka.binder.zkJaas.options::
+  Map with a key/value pair containing the login module options.
++
+Default: Empty map.
+
+Here is an example of launching a Spring Cloud Stream application with SASL and Kerberos using Spring Boot configuration properties:
+
+[source]
+----
+ java -Djava.security.auth.login.config=/path.to/kafka_client_jaas.conf -jar log.jar \
```
Comment (jvalkeal) [inline spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc]: I have a feeling that in your tests you've still used `-Djava.security.auth.login.config` as using below throws error:
```
java -jar time-source-kafka-1.0.3.BUILD-SNAPSHOT.jar \
  --spring.cloud.stream.kafka.binder.brokers=sandbox:6667 \
  --spring.cloud.stream.kafka.binder.zkNodes=sandbox:2181 \
  --spring.cloud.stream.bindings.output.destination=ticktock1 \
  --spring.cloud.stream.kafka.binder.configuration.security.protocol=SASL_PLAINTEXT \
  --spring.cloud.stream.kafka.binder.jaas.options.useKeyTab=true \
  --spring.cloud.stream.kafka.binder.jaas.options.storeKey=true \
  --spring.cloud.stream.kafka.binder.jaas.options.keyTab=/etc/security/keytabs/scdf.service.keytab \
  --spring.cloud.stream.kafka.binder.jaas.options.principal=scdf/sandbox.hortonworks.com@HORTONWORKS.COM \
  --spring.cloud.stream.kafka.binder.zkJaas.options.useKeyTab=true \
  --spring.cloud.stream.kafka.binder.zkJaas.options.storeKey=true \
  --spring.cloud.stream.kafka.binder.zkJaas.options.keyTab=/etc/security/keytabs/scdf.service.keytab \
  --spring.cloud.stream.kafka.binder.zkJaas.options.principal=scdf/sandbox.hortonworks.com@HORTONWORKS.COM \
  --server.port=0
```
```
Caused by: org.springframework.cloud.stream.binder.BinderException: Cannot initialize binder:
	at org.springframework.cloud.stream.binder.kafka.KafkaMessageChannelBinder.getPartitionsForTopic(KafkaMessageChannelBinder.java:524)
	at org.springframework.cloud.stream.binder.kafka.KafkaMessageChannelBinder.createProducerDestinationIfNecessary(KafkaMessageChannelBinder.java:230)
	at org.springframework.cloud.stream.binder.kafka.KafkaMessageChannelBinder.createProducerDestinationIfNecessary(KafkaMessageChannelBinder.java:93)
	at org.springframework.cloud.stream.binder.AbstractMessageChannelBinder.doBindProducer(AbstractMessageChannelBinder.java:98)
	at org.springframework.cloud.stream.binder.AbstractMessageChannelBinder.doBindProducer(AbstractMessageChannelBinder.java:53)
	at org.springframework.cloud.stream.binder.AbstractBinder.bindProducer(AbstractBinder.java:148)
	at org.springframework.cloud.stream.binding.ChannelBindingService.bindProducer(ChannelBindingService.java:113)
	at org.springframework.cloud.stream.binding.BindableProxyFactory.bindOutputs(BindableProxyFactory.java:208)
	at org.springframework.cloud.stream.binding.OutputBindingLifecycle.start(OutputBindingLifecycle.java:57)
	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:173)
	... 20 more
Caused by: org.apache.kafka.common.KafkaException: Failed to construct kafka producer
	at org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:321)
	at org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:171)
	at org.springframework.kafka.core.DefaultKafkaProducerFactory.createKafkaProducer(DefaultKafkaProducerFactory.java:138)
	at org.springframework.kafka.core.DefaultKafkaProducerFactory.createProducer(DefaultKafkaProducerFactory.java:130)
	at org.springframework.cloud.stream.binder.kafka.KafkaMessageChannelBinder$4.doWithRetry(KafkaMessageChannelBinder.java:510)
	at org.springframework.cloud.stream.binder.kafka.KafkaMessageChannelBinder$4.doWithRetry(KafkaMessageChannelBinder.java:503)
	at org.springframework.retry.support.RetryTemplate.doExecute(RetryTemplate.java:263)
	at org.springframework.retry.support.RetryTemplate.execute(RetryTemplate.java:154)
	at org.springframework.cloud.stream.binder.kafka.KafkaMessageChannelBinder.getPartitionsForTopic(KafkaMessageChannelBinder.java:503)
	... 29 more
Caused by: org.apache.kafka.common.KafkaException: java.lang.IllegalArgumentException: You must pass java.security.auth.login.config in secure mode.
	at org.apache.kafka.common.network.SaslChannelBuilder.configure(SaslChannelBuilder.java:74)
	at org.apache.kafka.common.network.ChannelBuilders.create(ChannelBuilders.java:60)
	at org.apache.kafka.clients.ClientUtils.createChannelBuilder(ClientUtils.java:79)
	at org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:271)
	... 37 more
Caused by: java.lang.IllegalArgumentException: You must pass java.security.auth.login.config in secure mode.
	at org.apache.kafka.common.security.kerberos.Login.login(Login.java:289)
	at org.apache.kafka.common.security.kerberos.Login.<init>(Login.java:104)
	at org.apache.kafka.common.security.kerberos.LoginManager.<init>(LoginManager.java:44)
	at org.apache.kafka.common.security.kerberos.LoginManager.acquireLoginManager(LoginManager.java:85)
	at org.apache.kafka.common.network.SaslChannelBuilder.configure(SaslChannelBuilder.java:55)
	... 40 more
```
as there is yet another property check at https://github.com/apache/kafka/blob/0.9.0/clients/src/main/java/org/apache/kafka/common/security/kerberos/Login.java#L287
```
    private synchronized LoginContext login(final String loginContextName) throws LoginException {
        String jaasConfigFile = System.getProperty(JaasUtils.JAVA_LOGIN_CONFIG_PARAM);
        if (jaasConfigFile == null) {
            throw new IllegalArgumentException("You must pass " + JaasUtils.JAVA_LOGIN_CONFIG_PARAM + " in secure mode.");
        }

```

File: spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc [hunk 2]
```diff
@@ -215,18 +215,120 @@ spring.cloud.stream.kafka.binder.configuration.security.protocol=SASL_SSL
...
+   --spring.cloud.stream.kafka.binder.jaas.options.storeKey=true \
+   --spring.cloud.stream.kafka.binder.jaas.options.keyTab=/etc/security/keytabs/kafka_client.keytab \
+   --spring.cloud.stream.kafka.binder.jaas.options.principal=kafka-client-1@EXAMPLE.COM
+----
+
+As an alternative to setting `spring.cloud.stream.kafka.binder.autoCreateTopics` you can simply remove the broker dependency from the application. See <<exclude-admin-utils>> for details.
+
+[NOTE]
+====
+Do not mix JAAS configuration files and Spring Boot properties in the same application.
```
Comment (jvalkeal) [inline spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc]: All snippets above actually mixes these ;)

Comment (mbogoevici) [inline spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc]: That was a copy/paste mistake.

File: spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc [hunk 3]
```diff
@@ -215,18 +215,120 @@ spring.cloud.stream.kafka.binder.configuration.security.protocol=SASL_SSL
...
+
+Here is an example of launching a Spring Cloud Stream application with SASL and Kerberos using Spring Boot configuration properties:
+
+[source]
+----
+ java -Djava.security.auth.login.config=/path.to/kafka_client_jaas.conf -jar log.jar \
+   --spring.cloud.stream.kafka.binder.brokers=secure.server:9092 \
+   --spring.cloud.stream.kafka.binder.zkNodes=secure.zookeeper:2181 \
+   --spring.cloud.stream.bindings.input.destination=stream.ticktock \
+   --spring.cloud.stream.kafka.binder.clientConfiguration.security.protocol=SASL_PLAINTEXT \
```
Comment (mbogoevici) [inline spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc]: Will fix.

File: spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc (hunk 1 above)
Comment (mbogoevici) [inline spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc]: Turns out that this is specific to 0.9 - 0.10 doesn't require the file. I suggest we generate the file for 0.9.

File: spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc [hunk 4]
```diff
@@ -270,11 +270,10 @@ Here is an example of launching a Spring Cloud Stream application with SASL and
...
 ----
- java -Djava.security.auth.login.config=/path.to/kafka_client_jaas.conf -jar log.jar \
-   --spring.cloud.stream.kafka.binder.brokers=secure.server:9092 \
+ java --spring.cloud.stream.kafka.binder.brokers=secure.server:9092 \
    --spring.cloud.stream.kafka.binder.zkNodes=secure.zookeeper:2181 \
    --spring.cloud.stream.bindings.input.destination=stream.ticktock \
-   --spring.cloud.stream.kafka.binder.clientConfiguration.security.protocol=SASL_PLAINTEXT \
+   --spring.cloud.stream.kafka.binder.configuration.security.protocol=SASL_PLAINTEXT \
    --spring.cloud.stream.kafka.binder.jaas.options.useKeyTab=true \
    --spring.cloud.stream.kafka.binder.jaas.options.storeKey=true \
```
Comment (jvalkeal) [inline spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc]: I had to add
```
--spring.cloud.stream.kafka.binder.jaas.options.serviceName=kafka
```
as there was error
```
Caused by: java.lang.IllegalArgumentException: No serviceName defined in either JAAS or Kafka config
	at org.apache.kafka.common.security.kerberos.LoginManager.getServiceName(LoginManager.java:63)
	at org.apache.kafka.common.security.kerberos.LoginManager.<init>(LoginManager.java:45)
	at org.apache.kafka.common.security.kerberos.LoginManager.acquireLoginManager(LoginManager.java:85)
	at org.apache.kafka.common.network.SaslChannelBuilder.configure(SaslChannelBuilder.java:55)
```

File: spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc [hunk 5]
```diff
@@ -309,12 +308,11 @@ If the topics required already exist on the broker, or will be created by an adm
 
 [source]
 ----
- java -Djava.security.auth.login.config=/path.to/kafka_client_jaas.conf -jar log.jar \
-   --spring.cloud.stream.kafka.binder.brokers=secure.server:9092 \
+ java --spring.cloud.stream.kafka.binder.brokers=secure.server:9092 \
    --spring.cloud.stream.kafka.binder.zkNodes=secure.zookeeper:2181 \
    --spring.cloud.stream.bindings.input.destination=stream.ticktock \
    --spring.cloud.stream.kafka.binder.autoCreateTopics=false \
```
Comment (jvalkeal) [inline spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc]: If I have a working topic I was able to run apps without manual jaas file. Trying to hook to a topic which would get created will not work. Eventually tried to add `--spring.cloud.stream.kafka.binder.zkJaas.options.serviceName=zookeeper` but it didn't help anything.

So it's not able to create a topic with a leader as seen from here:
```
[root@sandbox bin]# ./kafka-topics.sh --zookeeper sandbox.hortonworks.com:2181 --describe --topic ticktock5
Topic:ticktock5	PartitionCount:1	ReplicationFactor:1	Configs:
	Topic: ticktock5	Partition: 0	Leader: none	Replicas: 1001	Isr: 
[root@sandbox bin]# ./kafka-topics.sh --zookeeper sandbox.hortonworks.com:2181 --describe --topic ticktock1
Topic:ticktock1	PartitionCount:1	ReplicationFactor:1	Configs:
	Topic: ticktock1	Partition: 0	Leader: 1001	Replicas: 1001	Isr: 1001
```
app just logged:
```
2016-09-20 13:22:47.247  WARN 1005 --- [ad | producer-1] org.apache.kafka.clients.NetworkClient   : Error while fetching metadata with correlation id 0 : {ticktock5=LEADER_NOT_AVAILABLE}
2016-09-20 13:22:47.445  WARN 1005 --- [ad | producer-1] org.apache.kafka.clients.NetworkClient   : Error while fetching metadata with correlation id 1 : {ticktock5=LEADER_NOT_AVAILABLE}
2016-09-20 13:22:47.547  WARN 1005 --- [ad | producer-1] org.apache.kafka.clients.NetworkClient   : Error while fetching metadata with correlation id 2 : {ticktock5=LEADER_NOT_AVAILABLE}
```

**Comment by jvalkeal** [comment]
Hmm, wait a second. Topic is created ok if I just use my jaas(don't have that zk Client section) file I generate in ambari.

```
[root@sandbox bin]# cat /etc/scdf/conf/scdf_kafka_jaas.conf 
KafkaClient {
   com.sun.security.auth.module.Krb5LoginModule required
   useKeyTab=true
   keyTab="/etc/security/keytabs/scdf.service.keytab"
   storeKey=true
   useTicketCache=false
   serviceName="kafka"
   principal="scdf/sandbox.hortonworks.com@HORTONWORKS.COM";
};
```
So if I remove `zkJaas` config, topic is created with a leader and everything works!

```
java -jar time-source-kafka-1.0.3.BUILD-SNAPSHOT.jar \
  --spring.cloud.stream.kafka.binder.brokers=sandbox:6667 \
  --spring.cloud.stream.kafka.binder.zkNodes=sandbox:2181 \
  --spring.cloud.stream.bindings.output.destination=ticktock6 \
  --spring.cloud.stream.kafka.binder.configuration.security.protocol=SASL_PLAINTEXT \
  --spring.cloud.stream.kafka.binder.jaas.options.useKeyTab=true \
  --spring.cloud.stream.kafka.binder.jaas.options.storeKey=true \
  --spring.cloud.stream.kafka.binder.jaas.options.serviceName=kafka \
  --spring.cloud.stream.kafka.binder.jaas.options.keyTab=/etc/security/keytabs/scdf.service.keytab \
  --spring.cloud.stream.kafka.binder.jaas.options.principal=scdf/sandbox.hortonworks.com@HORTONWORKS.COM \
  --server.port=0

java -jar log-sink-kafka-1.0.3.BUILD-SNAPSHOT.jar \
  --spring.cloud.stream.kafka.binder.brokers=sandbox:6667 \
  --spring.cloud.stream.kafka.binder.zkNodes=sandbox:2181 \
  --spring.cloud.stream.bindings.input.destination=ticktock6 \
  --spring.cloud.stream.kafka.binder.configuration.security.protocol=SASL_PLAINTEXT \
  --spring.cloud.stream.kafka.binder.jaas.options.useKeyTab=true \
  --spring.cloud.stream.kafka.binder.jaas.options.storeKey=true \
  --spring.cloud.stream.kafka.binder.jaas.options.serviceName=kafka \
  --spring.cloud.stream.kafka.binder.jaas.options.keyTab=/etc/security/keytabs/scdf.service.keytab \
  --spring.cloud.stream.kafka.binder.jaas.options.principal=scdf/sandbox.hortonworks.com@HORTONWORKS.COM \
  --server.port=0

[root@sandbox bin]# ./kafka-topics.sh --zookeeper sandbox.hortonworks.com:2181 --describe --topic ticktock6
Topic:ticktock6	PartitionCount:1	ReplicationFactor:1	Configs:
	Topic: ticktock6	Partition: 0	Leader: 1001	Replicas: 1001	Isr: 1001
```

File: spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc (hunk 4 above)
Comment (mbogoevici) [inline spring-cloud-stream-binder-kafka-docs/src/main/asciidoc/overview.adoc]: Ok, this can be prepopulated in the options, no need to add it manually.

**Comment by jvalkeal** [comment]
lgtm, works with and without jaas file.

**Comment by mbogoevici** [comment]
Merged via https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/commit/30c868e51b6d563d8488948476947c50997da4ff

**Pull request closed without merging**